src/app/features/courses/components/courses-list/components/courses-search/courses-search.component.ts

```diff
@@ -1,21 +1,47 @@
-import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
+import {
+  Component,
+  EventEmitter,
+  OnDestroy,
+  OnInit,
+  Output,
+} from '@angular/core';
+import { fromEvent, map, filter, debounceTime, Subscription } from 'rxjs';
 
 @Component({
   selector: 'app-courses-search',
   templateUrl: './courses-search.component.html',
   styleUrls: ['./courses-search.component.scss'],
 })
-export class CoursesSearchComponent implements OnInit {
-  searchValue: string = '';
+export class CoursesSearchComponent implements OnInit, OnDestroy {
+  private searchElement!: HTMLInputElement | null;
+  private subs: Subscription[] = [];
 
   @Output()
   searchValueChange = new EventEmitter<string>();
 
   constructor() {}
 
-  ngOnInit(): void {}
+  ngOnInit(): void {
+    this.searchElement = document.querySelector('.search__input');
 
-  onSearch() {
-    this.searchValueChange.emit(this.searchValue);
+    if (this.searchElement) {
+      const searchEvent = fromEvent(this.searchElement, 'keyup').pipe(
+        map((e) => (e.target as HTMLInputElement).value),
+        filter((text) => text.length > 2 || text.length === 0),
+        debounceTime(1000)
+      );
+
+      this.subs.push(
```

Comment:
This array with subs looks not great at all. There is a better way to unsubscribe with `takeUntil `operator. Here is a code example:

```
  private readonly destroy$ = new Subject<boolean>();

  public flights: FlightModel[];

  constructor(private readonly flightService: FlightService) {}

  ngOnInit(): void {
    this.flightService
      .getAll()
      .pipe(takeUntil(this.destroy$))
      .subscribe(flights => (this.flights = flights));
  }

  ngOnDestroy(): void {
    this.destroy$.next(true);
    this.destroy$.complete();
  }
```

Reply:
done